File: components/auth/src/main/java/org/cloudfoundry/credhub/auth/X509AuthenticationFailureHandler.java
package org.cloudfoundry.credhub.auth;

import java.io.IOException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.cloudfoundry.credhub.ErrorMessages;
import org.cloudfoundry.credhub.views.ResponseError;

import static org.cloudfoundry.credhub.auth.X509AuthenticationProvider.CLIENT_AUTH_EXTENDED_KEY_USAGE;
import static org.springframework.util.MimeTypeUtils.APPLICATION_JSON;

@Component
public class X509AuthenticationFailureHandler implements AuthenticationFailureHandler {
  private static final String INVALID_DN_MESSAGE = "No matching pattern was found in subjectDN";
  private static final String INVALID_MTLS_ID_RESPONSE = ErrorMessages.Auth.INVALID_MTLS_IDENTITY;
  private static final String INVALID_CLIENT_AUTH_RESPONSE = ErrorMessages.Auth.MTLS_NOT_CLIENT_AUTH;

  private final ObjectMapper objectMapper;

  @Autowired
  X509AuthenticationFailureHandler(
    final ObjectMapper objectMapper
  ) {
    super();
    this.objectMapper = objectMapper;
  }

  @Override
  public void onAuthenticationFailure(
    final HttpServletRequest request,
    final HttpServletResponse response,
    final AuthenticationException exception
  ) throws IOException {
    if (exception.getMessage().contains(INVALID_DN_MESSAGE)) {
      writeUnauthorizedResponse(response, INVALID_MTLS_ID_RESPONSE);
    }

    if (exception.getMessage().contains("Certificate does not contain: " + CLIENT_AUTH_EXTENDED_KEY_USAGE)) {
      writeUnauthorizedResponse(response, INVALID_CLIENT_AUTH_RESPONSE);
    }
  }

  private void writeUnauthorizedResponse(final HttpServletResponse response, final String message) throws IOException {
    final ResponseError responseError = new ResponseError(message);

    response.setStatus(HttpStatus.UNAUTHORIZED.value());
    response.setContentType(APPLICATION_JSON.getType());
    response.getWriter().write(objectMapper.writeValueAsString(responseError));
  }
}

File: components/auth/src/main/kotlin/org/cloudfoundry/credhub/auth/X509AuthenticationFailureHandler.kt
package org.cloudfoundry.credhub.auth

import com.fasterxml.jackson.databind.ObjectMapper
import org.cloudfoundry.credhub.ErrorMessages
import org.cloudfoundry.credhub.auth.X509AuthenticationProvider.Companion.CLIENT_AUTH_EXTENDED_KEY_USAGE
import org.cloudfoundry.credhub.views.ResponseError
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.stereotype.Component
import org.springframework.util.MimeTypeUtils.APPLICATION_JSON
import java.io.IOException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse

@Component
class X509AuthenticationFailureHandler
    @Autowired
    internal constructor(
        private val objectMapper: ObjectMapper,
    ) : AuthenticationFailureHandler {
        @Throws(IOException::class)
        override fun onAuthenticationFailure(
            request: HttpServletRequest,
            response: HttpServletResponse,
            exception: AuthenticationException,
        ) {
            if (exception.message.toString().contains(INVALID_DN_MESSAGE)) {
                writeUnauthorizedResponse(response, INVALID_MTLS_ID_RESPONSE)
            }

            if (exception.message.toString().contains("Certificate does not contain: $CLIENT_AUTH_EXTENDED_KEY_USAGE")) {
                writeUnauthorizedResponse(response, INVALID_CLIENT_AUTH_RESPONSE)
            }
        }

        @Throws(IOException::class)
        private fun writeUnauthorizedResponse(
            response: HttpServletResponse,
            message: String,
        ) {
            val responseError = ResponseError(message)

            response.status = HttpStatus.UNAUTHORIZED.value()
            response.contentType = APPLICATION_JSON.type
            response.writer.write(objectMapper.writeValueAsString(responseError))
        }

        companion object {
            private const val INVALID_DN_MESSAGE = "No matching pattern was found in subjectDN"
            private const val INVALID_MTLS_ID_RESPONSE = ErrorMessages.Auth.INVALID_MTLS_IDENTITY
            private const val INVALID_CLIENT_AUTH_RESPONSE = ErrorMessages.Auth.MTLS_NOT_CLIENT_AUTH
        }
    }
